// quickgame-cli/lib/commands/compile.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _webpack = _interopRequireDefault(require("webpack"));

var _webpackConf = _interopRequireDefault(require("../webpack-conf"));

module.exports = function () {
  var options = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  var mode = arguments.length > 1 ? arguments[1] : undefined;
  var webpackMode = mode === 'prod' ? 'production' : 'development';
  var config = (0, _webpackConf["default"])(options, webpackMode);
  (0, _webpack["default"])(config.webpackConf, function (err, stats) {
    if (err) {
      console.log(err);
      return;
    }

    var info = stats.toString({
      builtAt: true,
      timings: false,
      hash: false,
      entrypoints: false,
      children: false,
      chunks: false,
      chunkModules: false,
      chunkOrigins: false,
      modules: true,
      version: false,
      assets: false,
      colors: true
    });

    if (stats.hasErrors()) {
      console.log(info);
    }
  });
};
